core-models/src/main/scala/com/mesosphere/usi/core/models/template/RunTemplate.scala

```diff
@@ -0,0 +1,129 @@
+      b: Mesos.TaskInfo.Builder,
+      matchedOffer: Mesos.Offer,
+      taskName: TaskName,
+      taskIdStrategy: CurriedPodTaskIdStrategy,
+      resources: Seq[Mesos.Resource]): Unit = {
+    if (b.hasTaskId) {
+      logger.error(
+        s"TaskInfo builder for ${taskIdStrategy.podId} / ${taskName} set the task ID but shouldn't! Value is ignored")
+    }
+    if (b.hasAgentId) {
+      logger.error(
+        s"TaskInfo builder for ${taskIdStrategy.podId} / ${taskName} set the agentId but shouldn't! Value is ignored")
+    }
+    if (b.getResourcesCount != 0) {
+      logger.error(s"TaskInfo builder for ${taskIdStrategy.podId} set resources but shouldn't! Value is ignored")
+      b.clearResources()
+    }
+    val taskId = taskIdStrategy(taskName).value
+    if (!b.hasName) {
+      b.setName(taskName.value)
+    }
+    b.setTaskId(Mesos.TaskID.newBuilder().setValue(taskId))
+    b.setAgentId(matchedOffer.getAgentId)
+    b.addAllResources(resources.asJava)
+  }
+}
+
+class LaunchGroupRunTemplate(val role: String, executorBuilder: ExecutorBuilder, tasks: Map[TaskName, TaskBuilder])
+    extends RunTemplate
+    with StrictLogging {
+  val taskResourceRequirements = tasks.map { case (taskId, taskBuilder) => taskId -> taskBuilder.resourceRequirements }
+
+  override val executorResources: Seq[ResourceRequirement] = Nil
+
+  override private[usi] def buildLaunchOperation(
+      matchedOffer: Mesos.Offer,
+      taskIdStrategy: CurriedPodTaskIdStrategy,
+      executorResources: Seq[Mesos.Resource],
+      taskResources: Map[TaskName, Seq[Mesos.Resource]]
+  ): Either[Mesos.Offer.Operation.Launch, Mesos.Offer.Operation.LaunchGroup] = {
+    val executorProtoBuilder = executorBuilder.buildExecutor(matchedOffer, executorResources)
+    if (executorProtoBuilder.getResourcesCount != 0) {
+      logger.error(s"Executor builder for ${taskIdStrategy.podId} set resources but shouldn't! Value is ignored")
+      executorProtoBuilder.clearResources()
+    }
+    executorProtoBuilder.addAllResources(executorResources.asJava)
+
+    val taskInfos = taskResources.map {
+      case (taskName, resources) =>
+        val b = tasks(taskName).buildTask(matchedOffer, resources, taskResources)
+        RunTemplate.setTaskInfo(b, matchedOffer, taskName, taskIdStrategy, resources)
+        taskName -> b.build
+    }
+
+    val tgBuilder = Mesos.TaskGroupInfo
+      .newBuilder()
+      .addAllTasks(taskInfos.values.asJava)
+
+    val launchGroup = Mesos.Offer.Operation.LaunchGroup
+      .newBuilder()
+      .setExecutor(executorProtoBuilder)
+      .setTaskGroup(tgBuilder)
+      .build()
+
+    Right(launchGroup)
+  }
+}
+
+class LegacyLaunchRunTemplate(val role: String, taskBuilder: TaskBuilder) extends RunTemplate {
+  final override val taskResourceRequirements: Map[TaskName, Seq[ResourceRequirement]] =
+    Map(TaskName.empty -> taskBuilder.resourceRequirements)
+
+  final override val executorResources: Seq[ResourceRequirement] = Nil
```

Comment:
Shouldn't this be an empty seq? In the code above we access with with .foreach without checking for Nil...

```suggestion
  final override val executorResources: Seq[ResourceRequirement] = Seq.empty
```

Reply:
Nil is List.empty which implements Seq